src/Bicep.Core/Semantics/Compilation.cs

```diff
@@ -49,27 +50,31 @@ public Compilation(IFeatureProviderFactory featureProviderFactory, INamespacePro
         public INamespaceProvider NamespaceProvider { get; }
 
         public SemanticModel GetEntrypointSemanticModel()
-            => GetSemanticModel(SourceFileGrouping.EntryPoint);
+            // entry point semantic models are guaranteed to cast successfully
+            => (SemanticModel)GetSemanticModel(SourceFileGrouping.EntryPoint);
 
         public SemanticModel GetSemanticModel(BicepFile bicepFile)
             => this.GetSemanticModel<SemanticModel>(bicepFile);
 
         public ArmTemplateSemanticModel GetSemanticModel(ArmTemplateFile armTemplateFile)
             => this.GetSemanticModel<ArmTemplateSemanticModel>(armTemplateFile);
 
+        public SemanticModel GetSemanticModel(BicepParamFile bicepParamFile)
```

Comment:
If we're confident that `IBicepSourceFile` will always give a `SemanticModel` (what your above comment is indicating), we could simplify things by having a single overload instead of separate ones for `BicepFile` & `BicepParamFile`:
`public SemanticModel GetSemanticModel(IBicepSourceFile sourceFIle)`

This would avoid the cast on line 54, and the ternary on line 71

Reply:
Yeah, all of this needs to be simplified. I think it's an artifact of when I was going to have separate semantic model types and then I never cleaned up when I changed my mind.